From 0a6b205fed327c4649c6204a2e8c3b12b29c8459 Mon Sep 17 00:00:00 2001
From: Thomas Huth <thuth@redhat.com>
Date: Wed, 1 Oct 2025 10:47:08 +0200
Subject: [PATCH] Revert "hw/s390x/s390-virtio-ccw: Remove the deprecated 4.0
 machine type"

JIRA: https://issues.redhat.com/browse/RHEL-105900
Upstream status: downstream-only

This reverts commit 1b432c51cd12e778079ee0cfd7e4c51aad928c65.

We still need the code for the old machine types since the -rhel7.6.0
machine type is based on these old upstream machine types.

Signed-off-by: Thomas Huth <thuth@redhat.com>
---
 hw/s390x/s390-virtio-ccw.c  | 14 ++++++++++++++
 target/s390x/gen-features.c |  4 ++++
 2 files changed, 18 insertions(+)

diff --git a/hw/s390x/s390-virtio-ccw.c b/hw/s390x/s390-virtio-ccw.c
index 0ff4114d2a..ff2efc1b93 100644
--- a/hw/s390x/s390-virtio-ccw.c
+++ b/hw/s390x/s390-virtio-ccw.c
@@ -1182,6 +1182,20 @@ static void ccw_machine_4_1_class_options(MachineClass *mc)
 }
 DEFINE_CCW_MACHINE(4, 1);
 
+static void ccw_machine_4_0_instance_options(MachineState *machine)
+{
+    static const S390FeatInit qemu_cpu_feat = { S390_FEAT_LIST_QEMU_V4_0 };
+    ccw_machine_4_1_instance_options(machine);
+    s390_set_qemu_cpu_model(0x2827, 12, 2, qemu_cpu_feat);
+}
+
+static void ccw_machine_4_0_class_options(MachineClass *mc)
+{
+    ccw_machine_4_1_class_options(mc);
+    compat_props_add(mc->compat_props, hw_compat_4_0, hw_compat_4_0_len);
+}
+DEFINE_CCW_MACHINE(4, 0);
+
 #endif /* disabled for RHEL */
 
 static void ccw_rhel_machine_9_6_0_instance_options(MachineState *machine)
diff --git a/target/s390x/gen-features.c b/target/s390x/gen-features.c
index a814ece82f..4346b92431 100644
--- a/target/s390x/gen-features.c
+++ b/target/s390x/gen-features.c
@@ -878,6 +878,9 @@ static uint16_t qemu_MIN[] = {
     S390_FEAT_ADAPTER_INT_SUPPRESSION,
     S390_FEAT_MSA_EXT_3,
     S390_FEAT_MSA_EXT_4,
+};
+
+static uint16_t qemu_V4_0[] = {
     /*
      * Only BFP bits are implemented (HFP, DFP, PFPO and DIVIDE TO INTEGER not
      * implemented yet).
@@ -1049,6 +1052,7 @@ static FeatGroupDefSpec FeatGroupDef[] = {
  *******************************/
 static FeatGroupDefSpec QemuFeatDef[] = {
     QEMU_FEAT_INITIALIZER(MIN),
+    QEMU_FEAT_INITIALIZER(V4_0),
     QEMU_FEAT_INITIALIZER(V4_1),
     QEMU_FEAT_INITIALIZER(V6_0),
     QEMU_FEAT_INITIALIZER(V6_2),
